File: typescript/oop-patterns/creational-patterns/abstract-factory/abstract-factory.ts
// интерфейс абстрактной фабрики делает совместимыми конкретные фабрики и стандартизирует
// типы продуктов, которые они выпускают
interface AbstractFactory {
  createProductA: () => AbstractProductA;
  createProductB: () => AbstractProductB;
}

// Абстрактный интерфейс для каждого выпускаемого типа продуктов стандартизирует
// признаки, присущие конкретному типу продукта, что делает их взаимозаменяемыми
interface AbstractProductA {
  mandatoryFeatureOfA: () => void;
}

interface AbstractProductB {
  mandatoryFeatureOfB: () => void;
}

// Конкретный продукт следует интерфейсу своего типа, но может иметь и свои особенные черты
class ConcreteProductAType1 implements AbstractProductA {
  mandatoryFeatureOfA() {
    console.log('Все продукты вида "А" имеют эту особенность');
  }
  specialFeatureOfAType1() {
    console.log('Эта особенность только у продуктов вида "А", типа 1');
  }
}

class ConcreteProductBType1 implements AbstractProductB {
  mandatoryFeatureOfB() {
    console.log('Все продукты вида "Б" имеют эту особенность');
  }
  specialFeatureOfBType1() {
    console.log('Эта особенность только у продуктов вида "Б", типа 1');
  }
}

class ConcreteProductAType2 implements AbstractProductA {
  mandatoryFeatureOfA() {
    console.log('Все продукты вида "А" имеют эту особенность');
  }
  specialFeatureOfAType2() {
    console.log('Эта особенность только у продуктов вида "А", типа 2');
  }
}

class ConcreteProductBType2 implements AbstractProductB {
  mandatoryFeatureOfB() {
    console.log('Все продукты вида "Б" имеют эту особенность');
  }
  specialFeatureOfBType2() {
    console.log('Эта особенность только у продуктов вида "Б", типа 2');
  }
}

// Конкретная фабрика, следуя интерфейсу абстрактной фабрики, выпускает все необходимые
// типы конкретных продуктов, при этом каждая фабрика производит ряд этих продуктов со
// своими особенностями (например, одна фабрика производит набор кнопок черного цвета, а другая
// тот же набор белого цвета)
class ConcreteFactory1 implements AbstractFactory {
  createProductA() {
    return new ConcreteProductAType1();
  }
  createProductB() {
    return new ConcreteProductBType1();
  }
}

class ConcreteFactory2 implements AbstractFactory {
  createProductA() {
    return new ConcreteProductAType2();
  }
  createProductB() {
    return new ConcreteProductBType2();
  }
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
const FactoryOfType1Products = new ConcreteFactory1();
const FactoryOfType2Products = new ConcreteFactory2();

const productAType1 = FactoryOfType1Products.createProductA();
const productAType2 = FactoryOfType2Products.createProductA();
const productBType1 = FactoryOfType1Products.createProductB();
const productBType2 = FactoryOfType2Products.createProductB();

productAType1.mandatoryFeatureOfA();
productAType1.specialFeatureOfAType1();
productAType2.mandatoryFeatureOfA();
productAType2.specialFeatureOfAType2();
productBType1.mandatoryFeatureOfB();
productBType1.specialFeatureOfBType1();
productBType2.mandatoryFeatureOfB();
productBType2.specialFeatureOfBType2();

File: go/oop-patterns/creational-patterns/abstract-factory/abstract-factory.go
package main

import "fmt"

// интерфейс абстрактной фабрики делает совместимыми конкретные фабрики и стандартизирует
// типы продуктов, которые они выпускают
type abstractFactory interface {
	createProductA() *abstractProductA
	createProductB() *abstractProductB
}

// интерфейс для каждого выпускаемого типа продуктов стандартизирует
// признаки, присущие конкретному типу продукта, что делает их взаимозаменяемыми
type abstractProductA interface {
	mandatoryFeatureOfA()
}

type abstractProductB interface {
	mandatoryFeatureOfB()
}

// Конкретный продукт следует интерфейсу своего типа, но может иметь и свои особенные черты
type concreteProductAType1 struct {
}

func (p *concreteProductAType1) mandatoryFeatureOfA() {
	fmt.Println("All A products have this feature")
}
func (p *concreteProductAType1) specialFeatureOfAType1() {
	fmt.Println("This feature is only in As of type 1")
}

type concreteProductAType2 struct {
}

func (p *concreteProductAType2) mandatoryFeatureOfA() {
	fmt.Println("All A products have this feature")
}
func (p *concreteProductAType2) specialFeatureOfAType2() {
	fmt.Println("This feature is only in As of type 2")
}

type concreteProductBType1 struct {
}

func (p *concreteProductBType1) mandatoryFeatureOfB() {
	fmt.Println("All B products have this feature")
}
func (p *concreteProductBType1) specialFeatureOfBType1() {
	fmt.Println("This feature is only in Bs of type 1")
}

type concreteProductBType2 struct {
}

func (p *concreteProductBType2) mandatoryFeatureOfB() {
	fmt.Println("All B products have this feature")
}
func (p *concreteProductBType2) specialFeatureOfBType2() {
	fmt.Println("This feature is only in Bs of type 2")
}

// Конкретная фабрика, следуя интерфейсу абстрактной фабрики, выпускает все необходимые
// типы конкретных продуктов, при этом каждая фабрика производит ряд этих продуктов со
// своими особенностями (например, одна фабрика производит набор кнопок черного цвета, а другая
// тот же набор белого цвета)
type concreteFactory1 struct {
}

func (f *concreteFactory1) createProductA() abstractProductA {
	return &concreteProductAType1{}
}
func (f *concreteFactory1) createProductB() abstractProductB {
	return &concreteProductBType1{}
}

type concreteFactory2 struct {
}

func (f *concreteFactory2) createProductA() abstractProductA {
	return &concreteProductAType2{}
}
func (f *concreteFactory2) createProductB() abstractProductB {
	return &concreteProductBType2{}
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// ------------------------------- Клиентский код -------------------------------
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
func main() {
	factoryOfType1Products := &concreteFactory1{}
	factoryOfType2Products := &concreteFactory2{}

	productAType1 := factoryOfType1Products.createProductA()
	productAType2 := factoryOfType2Products.createProductA()
	productBType1 := factoryOfType1Products.createProductB()
	productBType2 := factoryOfType2Products.createProductB()

	productAType1.mandatoryFeatureOfA()
	productAType1.(*concreteProductAType1).specialFeatureOfAType1()
	productAType2.mandatoryFeatureOfA()
	productAType2.(*concreteProductAType2).specialFeatureOfAType2()

	productBType1.mandatoryFeatureOfB()
	productBType1.(*concreteProductBType1).specialFeatureOfBType1()
	productBType2.mandatoryFeatureOfB()
	productBType2.(*concreteProductBType2).specialFeatureOfBType2()
}
